scico/optimize/_admm.py

```diff
@@ -0,0 +1,390 @@
+        self,
+        f: Functional,
+        g_list: List[Functional],
+        C_list: List[LinearOperator],
+        rho_list: List[float],
+        alpha: float = 1.0,
+        x0: Optional[Union[JaxArray, BlockArray]] = None,
+        maxiter: int = 100,
+        subproblem_solver: Optional[SubproblemSolver] = None,
+        itstat_options: Optional[dict] = None,
+    ):
+        r"""Initialize an :class:`ADMM` object.
+
+        Args:
+            f: Functional :math:`f` (usually a loss function).
+            g_list: List of :math:`g_i` functionals. Must be same length
+                 as :code:`C_list` and :code:`rho_list`.
+            C_list: List of :math:`C_i` operators.
+            rho_list: List of :math:`\rho_i` penalty parameters.
+                Must be same length as :code:`C_list` and :code:`g_list`.
+            alpha: Relaxation parameter. No relaxation for default 1.0.
+            x0: Initial value for :math:`\mb{x}`. If ``None``, defaults
+                to an array of zeros.
+            maxiter: Number of ADMM outer-loop iterations. Default: 100.
+            subproblem_solver: Solver for :math:`\mb{x}`-update step.
+                Defaults to ``None``, which implies use of an instance of
+                :class:`GenericSubproblemSolver`.
+            itstat_options: A dict of named parameters to be passed to
+                the :class:`.diagnostics.IterationStats` initializer. The
+                dict may also include an additional key "itstat_func"
+                with the corresponding value being a function with two
+                parameters, an integer and an `ADMM` object, responsible
+                for constructing a tuple ready for insertion into the
+                :class:`.diagnostics.IterationStats` object. If ``None``,
+                default values are used for the dict entries, otherwise
+                the default dict is updated with the dict specified by
+                this parameter.
+        """
+        N = len(g_list)
+        if len(C_list) != N:
+            raise Exception(f"len(C_list)={len(C_list)} not equal to len(g_list)={N}")
+        if len(rho_list) != N:
+            raise Exception(f"len(rho_list)={len(rho_list)} not equal to len(g_list)={N}")
+
+        self.f: Functional = f
+        self.g_list: List[Functional] = g_list
+        self.C_list: List[LinearOperator] = C_list
+        self.rho_list: List[float] = rho_list
+        self.alpha: float = alpha
+        self.itnum: int = 0
+        self.maxiter: int = maxiter
+        self.timer: Timer = Timer()
+        if subproblem_solver is None:
+            subproblem_solver = GenericSubproblemSolver()
+        self.subproblem_solver: SubproblemSolver = subproblem_solver
+        self.subproblem_solver.internal_init(self)
+
+        # iteration number and time fields
+        itstat_fields = {
+            "Iter": "%d",
+            "Time": "%8.2e",
+        }
+        itstat_attrib = ["itnum", "timer.elapsed()"]
```

Comment:
This itstat initialization is always a stumbling block for me when reading the class. Could it be easily refactored into its own function/class?

Reply:
Perhaps. It would probably have to be some sort of private auxiliary class rather than a general utility, but that's worth considering. I'll make a separate issue as a reminder.